scipy/sparse/linalg/isolve/iterative.py

```diff
@@ -814,3 +811,190 @@ def id(b):
         info = iter_
 
     return postprocess(x), info
+
+
+#======================
+#   CGNE
+#======================
+@set_docstring('Use Conjugate Gradient Normal Error iteration to solve '
+               '``Ax = b``.',
+               'The real or complex N-by-N matrix of the linear system.\n'
+               'Alternatively, ``A`` can be a linear operator which can\n'
+               'produce ``Ax`` using, e.g.,\n'
+               '``scipy.sparse.linalg.LinearOperator``.')
+@non_reentrant()
+def cgne(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None):
+    """
+    References
+    ----------
+    [1] Yousef Saad, Iterative Methods for Sparse Linear Systems, Second Edition, SIAM, 2003
+    [2] H.Elman, Iterative methods for large, sparse, nonsymmetric systems of linear equations, R        esearch Report 229, Yale University (1982)
+    """
+    assert A.shape[0] == A.shape[1] and A.shape[0] == len(b), "The size of the matrix and the right-hand side does not match."
+    if maxiter is None:
+        maxiter = 10000
+    # type judgment and conversion
+    dtype = A.dtype
+    if dtype == int:
+        dtype = float
+        A = A.astype(dtype)
+        if b.dtype == int:
+            b = b.astype(dtype)
+    A, M, x, b, postprocess = make_system(A, M, x0, b)
+    # judge if b is a zero vector
+    if np.linalg.norm(b) == 0.:
+        x = b.copy()
+        return (postprocess(x), 0)
+    r = b - A.matvec(x)
+    At = A.H
+    z = M.matvec(r)
+    p = At.matvec(z)
+    ztr_old = np.inner(z.conjugate(), r)
+    r0norm = np.linalg.norm(r)
+    if r0norm == 0.:
+        return (postprocess(x), 0)
+    for iter in range(maxiter):
+        w = A.matvec(p)
+        alpha = ztr_old / np.inner(p.conjugate(), p)
+        x += alpha * p
+        r -= alpha * w
+        alpha = np.linalg.norm(r) / r0norm
+        if callback is not None:
+            callback(x)
+        if alpha < tol:
+            return (postprocess(x), 0)
+        elif iter == maxiter - 1:
+            return (postprocess(x), maxiter)
+        z = M.matvec(r)
+        ztr = np.inner(z.conjugate(), r)
+        beta = ztr / ztr_old
+        p = At.matvec(z) + beta * p
+        ztr_old = ztr
+
+
+#======================
+#   CGNR
+#======================
+@set_docstring('Use Conjugate Gradient Normal Residual iteration to solve '
```

Comment:
Seems like there is quite a bit of code duplication among the `cg` variants. Im wondering if using a unified function for all of them is better. Then access each particular variant through a keyword argument using `cg`? something like `cg(..., method='nr')` or similar. What do you think @rgommers

Reply:
Dear @zoj613 , @rgommers 
For CGNR， it is not common for most numerical software that using `cg` as the function name and `method=nr` as the keyword. Although CGNR is a variant of CG, please note that pepole don't want to split CGNR into `cg` and `nr` because people think it is a method for symmetric positive definite problem (SPD) when it comes to CG. But it means that one will solve asymmetric problem when it comes to CGNR (both `cg` and `nr` are not separated). Maybe it's a good way to deal with it because it is simple, but no one is willing to split the entire function name `cgnr` (CGNR). Thanks

Reply:
Okay, I see. Ignore my previous suggestion then.

Reply:
Dear @zoj613 ,
I'll submit the other PR with a faster GMRES later (For GMRES, your suggestion has been accepted). Because GMRES is relatively complicated, the code would be submitted individually to the convenience of review and comparison this changes. Thanks.